Changing a syscall API breaks the incremental build

**Describe the bug**
If a syscall interface changes (for example: adding a new parameter to an existing syscall) the generated syscall headers do not get updated.
This only applies while using Ninja as build system. It works with "Unix Makefile".

Testing under Ubuntu 24.04.

**To Reproduce**
Steps to reproduce the behavior:
1. build any sample application for any platform:
    `west build zephyr/samples/basic/blinky -b native_sim -- -G Ninja`
2. modify a syscall: for example add a new parameter to the gpio API.
    `patch include/zephyr/drivers/gpio.h gpio.patch`
3. re-build the application (no "pristine" build")
    `west build zephyr/samples/basic/blinky -b native_sim`
4. Build fails:
<details>
<summary>Error logs</summary>
```
[3/9] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj 
ccache /usr/bin/gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -D__ZEPHYR__=1 -I/home/david/zephyrproject/build/zephyr/include/generated/zephyr -I/home/david/zephyrproject/zephyr/include -I/home/david/zephyrproject/build/zephyr/include/generated -I/home/david/zephyrproject/zephyr/soc/native/inf_clock -I/home/david/zephyrproject/zephyr/boards/native/native_sim -I/home/david/zephyrproject/zephyr/scripts/native_simulator/common/src/include -I/home/david/zephyrproject/zephyr/scripts/native_simulator/native/src/include -I/home/david/zephyrproject/custom/cannectivity/include -fno-strict-aliasing -Os -imacros /home/david/zephyrproject/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/david/zephyrproject/zephyr/samples/basic/blinky=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/david/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/david/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -msse2 -mfpmath=sse -fvisibility=hidden -fno-freestanding -std=c11 -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /home/david/zephyrproject/zephyr/samples/basic/blinky/src/main.c
In file included from /home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:1908,
                 from /home/david/zephyrproject/zephyr/samples/basic/blinky/src/main.c:9:
/home/david/zephyrproject/build/zephyr/include/generated/zephyr/syscalls/gpio.h:23:12: error: conflicting types for ‘z_impl_gpio_pin_interrupt_configure’; have ‘int(const struct device *, gpio_pin_t,  gpio_flags_t)’ {aka ‘int(const struct device *, unsigned char,  unsigned int)’}
   23 | extern int z_impl_gpio_pin_interrupt_configure(const struct device * port, gpio_pin_t pin, gpio_flags_t flags);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:871:19: note: previous definition of ‘z_impl_gpio_pin_interrupt_configure’ with type ‘int(const struct device *, gpio_pin_t,  gpio_flags_t,  void *)’ {aka ‘int(const struct device *, unsigned char,  unsigned int,  void *)’}
  871 | static inline int z_impl_gpio_pin_interrupt_configure(const struct device *port,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/build/zephyr/include/generated/zephyr/syscalls/gpio.h:26:19: error: conflicting types for ‘gpio_pin_interrupt_configure’; have ‘int(const struct device *, gpio_pin_t,  gpio_flags_t)’ {aka ‘int(const struct device *, unsigned char,  unsigned int)’}
   26 | static inline int gpio_pin_interrupt_configure(const struct device * port, gpio_pin_t pin, gpio_flags_t flags)
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:866:15: note: previous declaration of ‘gpio_pin_interrupt_configure’ with type ‘int(const struct device *, gpio_pin_t,  gpio_flags_t,  void *)’ {aka ‘int(const struct device *, unsigned char,  unsigned int,  void *)’}
  866 | __syscall int gpio_pin_interrupt_configure(const struct device *port,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:866:15: warning: ‘gpio_pin_interrupt_configure’ used but never defined
[4/9] Building C object zephyr/drivers/gpio/CMakeFiles/drivers__gpio.dir/gpio_emul.c.obj
FAILED: zephyr/drivers/gpio/CMakeFiles/drivers__gpio.dir/gpio_emul.c.obj 
ccache /usr/bin/gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/david/zephyrproject/build/zephyr/include/generated/zephyr -I/home/david/zephyrproject/zephyr/include -I/home/david/zephyrproject/build/zephyr/include/generated -I/home/david/zephyrproject/zephyr/soc/native/inf_clock -I/home/david/zephyrproject/zephyr/boards/native/native_sim -I/home/david/zephyrproject/zephyr/scripts/native_simulator/common/src/include -I/home/david/zephyrproject/zephyr/scripts/native_simulator/native/src/include -I/home/david/zephyrproject/custom/cannectivity/include -Wshadow -fno-strict-aliasing -Os -imacros /home/david/zephyrproject/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/david/zephyrproject/zephyr/samples/basic/blinky=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/david/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/david/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -msse2 -mfpmath=sse -fvisibility=hidden -fno-freestanding -std=c11 -MD -MT zephyr/drivers/gpio/CMakeFiles/drivers__gpio.dir/gpio_emul.c.obj -MF zephyr/drivers/gpio/CMakeFiles/drivers__gpio.dir/gpio_emul.c.obj.d -o zephyr/drivers/gpio/CMakeFiles/drivers__gpio.dir/gpio_emul.c.obj -c /home/david/zephyrproject/zephyr/drivers/gpio/gpio_emul.c
In file included from /home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:1908,
                 from /home/david/zephyrproject/zephyr/drivers/gpio/gpio_emul.c:10:
/home/david/zephyrproject/build/zephyr/include/generated/zephyr/syscalls/gpio.h:23:12: error: conflicting types for ‘z_impl_gpio_pin_interrupt_configure’; have ‘int(const struct device *, gpio_pin_t,  gpio_flags_t)’ {aka ‘int(const struct device *, unsigned char,  unsigned int)’}
   23 | extern int z_impl_gpio_pin_interrupt_configure(const struct device * port, gpio_pin_t pin, gpio_flags_t flags);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:871:19: note: previous definition of ‘z_impl_gpio_pin_interrupt_configure’ with type ‘int(const struct device *, gpio_pin_t,  gpio_flags_t,  void *)’ {aka ‘int(const struct device *, unsigned char,  unsigned int,  void *)’}
  871 | static inline int z_impl_gpio_pin_interrupt_configure(const struct device *port,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/build/zephyr/include/generated/zephyr/syscalls/gpio.h:26:19: error: conflicting types for ‘gpio_pin_interrupt_configure’; have ‘int(const struct device *, gpio_pin_t,  gpio_flags_t)’ {aka ‘int(const struct device *, unsigned char,  unsigned int)’}
   26 | static inline int gpio_pin_interrupt_configure(const struct device * port, gpio_pin_t pin, gpio_flags_t flags)
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:866:15: note: previous declaration of ‘gpio_pin_interrupt_configure’ with type ‘int(const struct device *, gpio_pin_t,  gpio_flags_t,  void *)’ {aka ‘int(const struct device *, unsigned char,  unsigned int,  void *)’}
  866 | __syscall int gpio_pin_interrupt_configure(const struct device *port,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/include/zephyr/drivers/gpio.h:866:15: warning: ‘gpio_pin_interrupt_configure’ used but never defined
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/david/zephyrproject/build
```
</details>

**Expected behavior**

A change to the header file should rebuild the syscall header files and build the application without error.

**Impact**

A clean build is required to build the application after a changed syscall.

**Environment (please complete the following information):**

 - OS: Linux (Ubuntu 24.04)
 - Toolchain:  Zephyr SDK (0.17.0)
 - latest main (08269963370be2188a7c1898066f0fd6a3c8cbeb)


** Additional **

Patch file:
```diff
diff --git a/include/zephyr/drivers/gpio.h b/include/zephyr/drivers/gpio.h
index 3eed8194726..f3cb17defa1 100644
--- a/include/zephyr/drivers/gpio.h
+++ b/include/zephyr/drivers/gpio.h
@@ -865,11 +865,13 @@ static inline bool gpio_is_ready_dt(const struct gpio_dt_spec *spec)
  */
 __syscall int gpio_pin_interrupt_configure(const struct device *port,
 					   gpio_pin_t pin,
-					   gpio_flags_t flags);
+					   gpio_flags_t flags,
+					   void *dummy);
 
 static inline int z_impl_gpio_pin_interrupt_configure(const struct device *port,
 						      gpio_pin_t pin,
-						      gpio_flags_t flags)
+						      gpio_flags_t flags,
+						      void *dummy)
 {
 	const struct gpio_driver_api *api =
 		(const struct gpio_driver_api *)port->api;
@@ -950,9 +952,9 @@ static inline int z_impl_gpio_pin_interrupt_configure(const struct device *port,
  * @return a value from gpio_pin_interrupt_configure()
  */
 static inline int gpio_pin_interrupt_configure_dt(const struct gpio_dt_spec *spec,
-						  gpio_flags_t flags)
+						  gpio_flags_t flags, void* dummy)
 {
-	return gpio_pin_interrupt_configure(spec->port, spec->pin, flags);
+	return gpio_pin_interrupt_configure(spec->port, spec->pin, flags, dummy);
 }
 
 /**
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -96,7 +96,6 @@ set(OFFSETS_H_TARGET           offsets_h)
 set(SYSCALL_LIST_H_TARGET      syscall_list_h_target)
 set(DRIVER_VALIDATION_H_TARGET driver_validation_h_target)
 set(KOBJ_TYPES_H_TARGET        kobj_types_h_target)
-set(PARSE_SYSCALLS_TARGET      parse_syscalls_target)
 set(DEVICE_API_LD_TARGET       device_api_ld_target)
 
 define_property(GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT BRIEF_DOCS " " FULL_DOCS " ")
@@ -834,7 +833,7 @@ add_custom_command(
   --file-list        ${syscalls_file_list_output}
   $<$<BOOL:${CONFIG_EMIT_ALL_SYSCALLS}>:--emit-all-syscalls>
   DEPENDS ${syscalls_subdirs_trigger} ${PARSE_SYSCALLS_HEADER_DEPENDS}
-          ${syscalls_file_list_output} ${syscalls_interface}
+          ${syscalls_file_list_output} syscalls_interface
   )
 
 # Make sure Picolibc is built before the rest of the system; there's no explicit
@@ -851,12 +850,6 @@ set_property(TARGET ${SYSCALL_LIST_H_TARGET}
              ADDITIONAL_CLEAN_FILES
              ${CMAKE_CURRENT_BINARY_DIR}/include/generated/zephyr/syscalls
 )
-
-add_custom_target(${PARSE_SYSCALLS_TARGET}
-  DEPENDS
-  ${syscalls_json}
-  ${struct_tags_json}
-  )
 
 # 64-bit systems do not require special handling of 64-bit system call
 # parameters or return values, indicate this to the system call boilerplate
@@ -900,7 +893,7 @@ add_custom_command(
   COMMAND
   ${LEGACY_SYSCALL_LIST_H_ARGS}
   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
-  DEPENDS ${PARSE_SYSCALLS_TARGET}
+  DEPENDS ${syscalls_json}
   )
 
 include(${ZEPHYR_BASE}/cmake/kobj.cmake)
@@ -911,9 +904,13 @@ gen_kobject_list(
   OUTPUTS ${DRV_VALIDATION}
   SCRIPT_ARGS --validation-output ${DRV_VALIDATION}
   INCLUDES ${struct_tags_json}
+  DEPENDS ${struct_tags_json}
   )
 
-gen_kobject_list_headers(INCLUDES ${struct_tags_json})
+gen_kobject_list_headers(
+  INCLUDES ${struct_tags_json}
+  DEPENDS ${struct_tags_json}
+  )
 
 # Generate sections for kernel device subsystems
 set(
@@ -1270,6 +1267,7 @@ if(CONFIG_USERSPACE)
     OUTPUT ${KOBJECT_PREBUILT_HASH_LIST}
     KERNEL_TARGET ${ZEPHYR_LINK_STAGE_EXECUTABLE}
     INCLUDES ${struct_tags_json}
+    DEPENDS ${struct_tags_json}
     )
 
   add_custom_command(
@@ -1466,6 +1464,7 @@ if(CONFIG_USERSPACE)
     OUTPUT ${KOBJECT_HASH_LIST}
     KERNEL_TARGET ${ZEPHYR_LINK_STAGE_EXECUTABLE}
     INCLUDES ${struct_tags_json}
+    DEPENDS ${struct_tags_json}
     )
 
   # Use gperf to generate C code (KOBJECT_HASH_OUTPUT_SRC_PRE) which implements a
@@ -2298,7 +2297,7 @@ if(CONFIG_LLEXT_EDK)
       ${SYSCALL_SPLIT_TIMEOUT_ARG}
     COMMAND ${CMAKE_COMMAND}
         -P ${ZEPHYR_BASE}/cmake/llext-edk.cmake
-    DEPENDS ${logical_target_for_zephyr_elf} build_info_yaml_saved
+    DEPENDS ${logical_target_for_zephyr_elf} ${syscalls_json} build_info_yaml_saved
     COMMAND_EXPAND_LISTS
   )
   add_custom_target(llext-edk DEPENDS ${llext_edk_file})

--- a/cmake/kobj.cmake
+++ b/cmake/kobj.cmake
@@ -27,7 +27,6 @@ function(gen_kobject_list)
     DEPENDS
       ${arg_DEPENDS}
       ${GEN_KOBJECT_LIST}
-      ${PARSE_SYSCALLS_TARGET}
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
     )
   add_custom_target(${arg_TARGET} DEPENDS ${arg_OUTPUTS})
@@ -39,7 +38,7 @@ function(gen_kobject_list_gperf)
   cmake_parse_arguments(PARSE_ARGV 0 arg
     ""
     "TARGET;OUTPUT;KERNEL_TARGET"
-    "INCLUDES"
+    "INCLUDES;DEPENDS"
   )
   gen_kobject_list(
     TARGET ${arg_TARGET}
@@ -48,7 +47,9 @@ function(gen_kobject_list_gperf)
       --kernel $<TARGET_FILE:${arg_KERNEL_TARGET}>
       --gperf-output ${arg_OUTPUT}
     INCLUDES ${arg_INCLUDES}
-    DEPENDS ${arg_KERNEL_TARGET}
+    DEPENDS
+      ${arg_DEPENDS}
+      ${arg_KERNEL_TARGET}
     )
 endfunction()
 
@@ -58,7 +59,7 @@ function(gen_kobject_list_headers)
   cmake_parse_arguments(PARSE_ARGV 0 arg
     ""
     "GEN_DIR_OUT_VAR"
-    "INCLUDES"
+    "INCLUDES;DEPENDS"
   )
   if (PROJECT_BINARY_DIR)
     set(gen_dir ${PROJECT_BINARY_DIR}/include/generated/zephyr)
@@ -80,6 +81,9 @@ function(gen_kobject_list_headers)
       --kobj-otype-output ${KOBJ_OTYPE}
       --kobj-size-output ${KOBJ_SIZE}
     INCLUDES ${arg_INCLUDES}
+    DEPENDS
+      ${arg_DEPENDS}
+      ${arg_KERNEL_TARGET}
   )
 
   if(arg_GEN_DIR_OUT_VAR)

